runtime/compiler/control/CompilationThread.cpp

```diff
@@ -10869,6 +11001,24 @@ TR::CompilationInfo::compilationEnd(J9VMThread * vmThread, TR::IlGeneratorMethod
             else
                {
                jitMethodTranslated(vmThread, method, startPC);
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+               if (jitConfig->javaVM->internalVMFunctions->isCheckpointAllowed(vmThread)
+                   && jitConfig->javaVM->internalVMFunctions->isDebugOnRestoreEnabled(vmThread)
+                   && (!compInfo->isCheckpointInProgress() || comp->getOption(TR_FullSpeedDebug)))
```

Comment:
I don't understand the || condition here. Shouldn't that be a && ?

Reply:
This condition checks if we should save this method for recompilation post-restore. If debug on restore is enabled and we're pre-checkpoint, then we should (almost) always queue a method for recompilation because if we compiled it pre-checkpoint, it would have been generated with FSD enabled. The exception is proactive compilation; those will be compiled pre-checkpoint but may not be done with FSD enabled, so they should not be recompiled. 

That said, when we start a compilation, `isCheckpointInProgress` may have been false, so the current compilation is not a proactive compilation, but at this point in the code, `isCheckpointInProgress` could now be true and so the current compilation **should** be recompiled (because it was generated with FSD enabled). So that's why there's an `||`, to distinguish between proactive compilation and a compilation that occurred across a checkpoint request.

However, proactive compilations support will be added later on, so I guess technically `(!compInfo->isCheckpointInProgress() || comp->getOption(TR_FullSpeedDebug))` should be moved out to that future PR.